### PR TITLE
Solar board work enables temperature readings

### PR DESCRIPTION
**Pull request by tedhavelka**
This pull request brings in a new simple sensor driver for the Oresat solar board's temperature sensors, part TMP101AN.  The driver code is based on the ina226.[ch] driver files.  A new worker_t thread has been created, modeled after the app_blinky.[ch] example code in oresat-firmware/src/f0/app_solar/source.

There may be some Oresat coding standards missed in this first pull request by Ted H.  Also, the temperature sensor config register and its bit-wise flags are defined in tmp101an.h, but are not yet put to use as default settings provide a proof of functioning sensors.  One shot measurement mode to be implemented for low power purposes in a future pull request.

### CODE REVIEW

**Comment by heliochronix** [comment]
Hello,

I'm going to push a merge with the latest solar_dev and some file organization changes in preparation for a code review. Then I'll start going through things and performing a review.

File: src/f0/app_solar/source/tmp101.h [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+ */
+#define TMP101_PATCH                        0
+/** @} */
+
+
+/**
+ * @name    TMP101 Register Addresses
+ * @{
+ */
+#define TMP101_REG_TEMPERATURE_READING      0x00U
```
Comment (heliochronix) [inline src/f0/app_solar/source/tmp101.h:48]: The register seems to just be named "Temperature" in the datasheet, so perhaps we can just define this as TMP101_REG_TEMPERATURE

File: src/f0/app_solar/main.c [hunk 2]
```diff
@@ -48,11 +49,16 @@ static void app_init(void)
...
     init_worker(&worker2, "Solar Application", solar_wa, sizeof(solar_wa), NORMALPRIO, solar, NULL, true);
+    init_worker(&worker3, "Read Temperature", read_temperature_wa, sizeof(read_temperature_wa), NORMALPRIO, read_temperature, NULL, true);
+
     reg_worker(&worker1);
     reg_worker(&worker2);
+    reg_worker(&worker3);
 
     /* Start up debug output */
     sdStart(&SD2, NULL);
+
```
Comment (heliochronix) [inline src/f0/app_solar/main.c:60]: Looks like old code here

File: src/f0/app_solar/source/tmp101.c [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+void tmp101Start(TMP101Driver *devp, const TMP101Config *config) {
+//    i2cbuf_t buf;
+
+    osalDbgCheck((devp != NULL) && (config != NULL));
+    osalDbgAssert((devp->state == TMP101_STOP) ||
+            (devp->state == TMP101_READY),
+            "tmp101Start(), invalid state");
+
+    devp->config = config;
+
```
Comment (heliochronix) [inline src/f0/app_solar/source/tmp101.c:128]: It does not look like this code properly starts the I2C drivers (i2cStart). Take a look at the INA226 code for the proper starting and stopping of the I2C driver from the respective Start and Stop functions in this driver.

File: src/f0/app_solar/source/tmp101.c [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+ *
+ * @api
+ */
+void tmp101Stop(TMP101Driver *devp) {
+//    i2cbuf_t buf;
+
+    osalDbgCheck(devp != NULL);
+    osalDbgAssert((devp->state == TMP101_STOP) || (devp->state == TMP101_READY),
+            "tmp101Stop(), invalid state");
+
```
Comment (heliochronix) [inline src/f0/app_solar/source/tmp101.c:151]: Same comment as in tmp101Start but regarding i2cStop.

File: src/f0/app_solar/source/tmp101.c [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    buffer_tx[0] = TMP101_REG_TEMPERATURE_READING;
+    buffer_tx[1] = 0;
+
+// Clear first two bytes of receive buffer, as temperature reading has 12 bits spanning two bytes:
+    buffer_rx[0] = 0;
+    buffer_rx[1] = 0;
+
+
+#ifdef DIAG_ON__TMP101AN
+    chprintf((BaseSequentialStream *) &SD2, "reading temperature sensor %02X . . .\r\n", sensor_addr);
```
Comment (heliochronix) [inline src/f0/app_solar/source/tmp101.c:194]: Drivers should not have any chprintf related stuff in it

File: src/f0/app_solar/source/tmp101.c [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+#endif
+
+    i2cAcquireBus(devp->config->i2cp);
```
Comment (heliochronix) [inline src/f0/app_solar/source/tmp101.c:197]: This code should conditionally be compiled based on INA226_SHARED_BUS. Check the INA226 driver for an example of the conditional compile checks.

File: src/f0/app_solar/source/tmp101.c [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    i2cStart(devp->config->i2cp, devp->config->i2ccfg);
+
+    i2c_result = i2cMasterTransmitTimeout(devp->config->i2cp, sensor_addr, buffer_tx, 1, buffer_rx, 2, TIME_INFINITE);
+    i2c_flags = i2cGetErrors(devp->config->i2cp);
+    i2cReleaseBus(devp->config->i2cp);
```
Comment (heliochronix) [inline src/f0/app_solar/source/tmp101.c:202]: This code should conditionally be compiled based on INA226_SHARED_BUS. Check the INA226 driver for an example of the conditional compile checks.

File: src/f0/app_solar/source/tmp101.c [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+
+#ifdef DIAG_ON__TMP101AN
+    chprintf((BaseSequentialStream *) &SD2, "reading temperature sensor %02X . . .\r\n", sensor_addr);
+#endif
+
+    i2cAcquireBus(devp->config->i2cp);
+    i2cStart(devp->config->i2cp, devp->config->i2ccfg);
+
+    i2c_result = i2cMasterTransmitTimeout(devp->config->i2cp, sensor_addr, buffer_tx, 1, buffer_rx, 2, TIME_INFINITE);
```
Comment (heliochronix) [inline src/f0/app_solar/source/tmp101.c:200]: You should be making use of your tmp101I2CReadRegister functions here. This whole section should be conditionally compiled based on TMP101_USE_I2C. Check INA226 driver for an example.

File: src/f0/app_solar/source/tmp101.c [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    i2cflags_t i2c_flags = 0;
+    uint8_t sensor_addr = devp->config->saddr;
+    uint8_t buffer_tx[2] = {0};
+
+// Values chosen per TMP101AN datasheet:
+    buffer_tx[0] = TMP101_REG_TEMPERATURE_READING;
+    buffer_tx[1] = 0;
+
+// Prepare for I2C transaction:
+    i2cAcquireBus(devp->config->i2cp);
```
Comment (heliochronix) [inline src/f0/app_solar/source/tmp101.c:242]: Same comments about conditional compilation based on TMP101_USE_I2C and TMP101_SHARED_BUS

File: src/f0/app_solar/source/tmp101.c [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+    i2cStart(devp->config->i2cp, devp->config->i2ccfg);
+
+    i2c_result = i2cMasterTransmitTimeout(devp->config->i2cp, sensor_addr, buffer_tx, 1, byte_array, 2, TIME_INFINITE);
```
Comment (heliochronix) [inline src/f0/app_solar/source/tmp101.c:245]: Same comment about using tmp101I2CReadRegister

**Review by heliochronix — CHANGES_REQUESTED** [review]
Hello,

I've done a very quick initial pass through some of this, and have requested a few changes. There are some other things I'd like to address later on, but this is a start.

File: src/f0/app_solar/source/tmp101.h (hunk 1 above)
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.h:48]: Renamed per request.

File: src/f0/app_solar/main.c (hunk 2 above)
Comment (tedhavelka) [reply src/f0/app_solar/main.c:60]: Removed as used only for dev' purposes.

File: src/f0/app_solar/source/tmp101.c (hunk 3 above)
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c:128]: Code restored here to model the pattern in INA226 driver.  That code now more clear to pull requester.

File: src/f0/app_solar/source/tmp101.c (hunk 4 above)
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c:160]: Code restored here to model the pattern in INA226 driver.

File: src/f0/app_solar/source/tmp101.c (hunk 5 above)
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c]: Good catch, apologies!  Removed.

File: src/f0/app_solar/source/tmp101.c (hunk 6 above)
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c:213]: Removed.

File: src/f0/app_solar/source/tmp101.c (hunk 7 above)
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c:218]: Routine removed as newer version in place and tested.

File: src/f0/app_solar/source/tmp101.c (hunk 8 above)
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c:216]: This line also part of removed routine referenced in previous comment.

File: src/f0/app_solar/source/tmp101.c (hunk 10 above)
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c]: Corrected, routine tmp101ReadRegister() now called here.

File: src/f0/app_solar/source/tmp101.h [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+/** @} */
+
+
+
+// ORESAT_TASK_001 - add temperature sensor driver:
+// 2021-02-08 MON - first steps temperature reading
+#define OPTION_SENSOR_01 (1)
+#define OPTION_SENSOR_02 (2)
+
+#define I2C_ADDR_SENSOR_01 (0x48)
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.h:76]: TMP101 sensor address defines moved from driver header file to solar.h.

File: src/f0/app_solar/source/tmp101.h [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+#define TMP101AN__CONFIGURE_FAULT_COUNT_TO_ALERT__F0   (1 << 3)
+#define TMP101AN__CONFIGURE_FAULT_COUNT_TO_ALERT__F1   (1 << 4)
+#define TMP101AN__CONFIGURE_CONVERTER_RESOLUTION__R0   (1 << 5)
+#define TMP101AN__CONFIGURE_CONVERTER_RESOLUTION__R1   (1 << 6)
+#define TMP101AN__CONFIGURE_ONE_SHOT__OS               (1 << 7)
+/** @} */
+
+
+
+// ORESAT_TASK_001 - add temperature sensor driver:
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.h:71]: Some of the bit-wise definitions will soon be used, to put the TMP101 devices into low power mode.  That capability is not yet implemented but is planned.

File: src/f0/app_solar/source/tmp101.c [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+
+    i2c_flags = i2cGetErrors(devp->config->i2cp);
+    i2cReleaseBus(devp->config->i2cp);
+
+// TO BE optional warning code to report I2C bus errors:
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c]: Removed.

File: src/f0/app_solar/source/tmp101.c [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+#endif
+
+    i2cAcquireBus(devp->config->i2cp);
+    i2cStart(devp->config->i2cp, devp->config->i2ccfg);
+
+    i2c_result = i2cMasterTransmitTimeout(devp->config->i2cp, sensor_addr, buffer_tx, 1, buffer_rx, 2, TIME_INFINITE);
+    i2c_flags = i2cGetErrors(devp->config->i2cp);
+    i2cReleaseBus(devp->config->i2cp);
+
+#ifdef DIAG_ON__TMP101AN
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c]: Removed.

File: src/f0/app_solar/source/tmp101.h [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+#endif
+
+/**
+ * @brief   TMP101 shared I2C switch.
+ * @details If set to @p TRUE the device acquires I2C bus ownership
+ *          on each transaction.
+ * @note    The default is @p FALSE. Requires I2C_USE_MUTUAL_EXCLUSION.
+ */
+#if !defined(TMP101_SHARED_I2C) || defined(__DOXYGEN__)
+#define TMP101_SHARED_I2C                   TRUE
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.h:105]: Changed to 'FALSE'.

File: src/f0/app_solar/source/tmp101.h [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+ * @brief   TMP101AN temperature sensor driver
+ *
+ * @addtogroup TMP101
+ * @ingroup ORESAT
+ * @{
+ */
+#ifndef _TMP101_H_
+#define _TMP101_H_
+
+#include <chprintf.h>
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.h:12]: Removed.

File: src/f0/app_solar/source/tmp101.c [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+ */
+
+#include "hal.h"
+#include "tmp101.h"
+
+/*===========================================================================*/
+/* Driver local definitions.                                                 */
+/*===========================================================================*/
+
+#define DIAG_ON__TMP101AN
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c]: Removed.

File: src/f0/app_solar/source/tmp101.h [hunk 18]
```diff
@@ -0,0 +1,234 @@
+/**
+ * @file    tmp101.h
+ * @brief   TMP101AN temperature sensor driver
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.h:3]: Corrected TMP101AN to TMP101.

File: src/f0/app_solar/source/tmp101.c [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+/*==========================================================================*/
+/* Interface implementation.                                                */
+/*==========================================================================*/
+
+static const struct TMP101VMT vmt_device = {
+    (size_t)0,
+};
+
+/*===========================================================================*/
+/* - SECTION - Driver exported functions.                                    */
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c:90]: Fixed to match ChibiOS header namings.

File: src/f0/app_solar/source/tmp101.c (hunk 9 above)
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c:237]: Fixed, I2C related conditionals now in place.

File: src/f0/app_solar/source/tmp101.c [hunk 20]
```diff
@@ -0,0 +1,261 @@
+/**
+ * @file    tmp101.c
+ * @brief   TMP101 Digital to Analog Converter.
```
[the remark this answers is not in the document]
Comment (tedhavelka) [reply src/f0/app_solar/source/tmp101.c]: Fixed.

**Comment by tedhavelka** [comment]
Addressed and made changes, corrections per the requests by Heliochronix.  Requesting a review of this amended code.

**Pull request closed without merging**